README.md
# covid19-testing-scripts

Scripts that process data from/into the git repo `covid19-testing-data`



## Installation

```
pip3 install pew
pew new biominers_covid19
pip3 install notion pandas Click requests_cache==0.5.2 seaborn==0.10.1 matplotlib==2.1.2 statsmodels==0.11.1 bokeh
```

where `notion` is the "Unofficial Python API client for Notion.so" from https://github.com/jamalex/notion-py/
also mentioned at https://www.notion.so/Notion-Hacks-27b92f71afcd4ae2ac9a4d14fef0ce47


It also helps to see CSVs in the terminal with vd:

```
pip3 install visidata
```

## Usage

Step 0: import notion tables

- This requires the notion token, as documented at https://github.com/jamalex/notion-py/
- The exact documentation from that page of how to get this is:

    Obtain the `token_v2` value by inspecting your browser cookies on a logged-in session on Notion.so"

- In Chrome, this means: More tools, Developer tools, Application, Cookies, notion.so, `token_v2` value
  - and save that into a txt file
- Also, in notion, you need to have your own workspace (other than the workspace in which the tables exist)
- Finally, note that if the l0 step is giving an "unauthorized" error, check if your token needs updating


Command to run step 0:

```
python3 main.py l0-importbiominers \
    path/to/notion/key.txt \
    ~/Development/gitlab.com/biominers/covid19-testing-data/l0-notion_tables/multiple-biominers-gitrepo.csv 
```

To display more verbose logging during the l0 step, check related notes in main.py


Step 1: import non-biominer tables and merge with biominer

(This includes ourworldindata.org, wikipedia, and worldometers)

```
rm demo_cache.sqlite
python3 main.py l1-importothers \
  ~/Development/gitlab.com/biominers/covid19-testing-data/
```

After the first run of l1, there will most likely see an error showing country/date pairs on which there is a dip in cumulative tests.
Some of these errors are automatically marked as "to be skipped" (check the `is_approved` flag in the displayed table).
Others need manual attention. In any case, re-run the l1 step a few times until all the automatically skipped rows
are no longer showing up in the table, as follows:

```
python3 main.py l1-importothers --skip-download ~/Development/gitlab.com/biominers/covid19-testing-data/
```

where we use `--skip-download` to avoid downloading JHU, OWID, etc

Check the related section `Troubleshooting / How to deal with l1 error "Found 46 triplets with dips in their cumulative tests."`


Step 2: merge with confirmed cases

(TODO add extracting the latestOnly file as well)

```
python3 main.py l2-mergetogether \
  ~/Development/gitlab.com/biominers/covid19-testing-data/
```

At this stage, there could be either an error saying that the JHU confirmed cases are decreasing.
These are resolved by identifying some rows with replacement values in the l2/t15.csv file.
Usually the replacement value would be the lower value being copied backwards, as if the previous number
was wrongly inflated. After this fix, a re-run of l2 is sufficient.

Another error is that tests is less than cases. In that case, the program prompts the user
to check l2/drop.csv.new and merge it with l2/drop.csv and mark some rows as skippable.
After this fix, a re-run of l1 is necessary (no need to download), eg

```
python3 main.py l1-importothers --skip-download ~/Development/gitlab.com/biominers/covid19-testing-data/
```

followed by re-run of l2. This might need to be re-iterated a few times since at each iteration,
the next-best source for each country/date pair is pushed forward and tested again.
It's important to see the number of errors going down and converging towards 0 with each iteration.



Step 3: interpolate and extrapolate

```
python3 main.py l3-generatearcdata \
  ~/Development/gitlab.com/biominers/covid19-testing-data/
```

Note: As of 2020-05-15(?), this step yields a warning `A value is trying to be set on a copy of a slice from a DataFrame` that needs to be fixed.


After running step 3, we currently need to open all the `ArcGIS/v2/*csv` files manually in excel/libreoffice and resave the csv to drop the `.0` suffixes of decimals.

Casting the corresponding dataframe column to int doesnt work because it contains NAs, which cannot coexist in int column in pandas


Step 4a: Notebook t11d: R, chisquared thresholds.

This is a jupyter notebook stored in `/gdrive/biominers/shadi/`.

Either run it in your web browser in Google colab,
or run the script stored locally in this repository (exported from the notebook as of 2020-06-16) as follows:

```
# Install packages in R
install.packages(c("data.table", "zoo"))# , "caret", "hexbin"))

# Execute script from bash
Rscript t11d_shadi_chisquared_testing_on_total_tests_r.R
```

- input:
  - `l2/interpolated.csv`
  - `gitlab/biominers/kaggle/shadi/ellipse_lib.R`
- output:
  - `l4/t11d-chisquared-history-v20200512.csv`
  - deprecated: `l2/t11d-chisquared-ranks.csv`


Step 4b: perform analysis and generate plots

- input: `l4/t11d-chisquared-history-v20200512.csv`
- output: `l4/chisquared-postprocessed.csv` and other l4 html files

```

python3 main.py l4-plots \
  ~/Development/gitlab.com/biominers/covid19-testing-data/ \
  www/
```


Step 5: generate latest country table

- input: `/ArcGIS/v2/t11c-confirmedtotalTests-latestOnly.csv`
- output: `www/t11c-country_latest_table.html`

```

python3 main.py l5-generatetable \
  ~/Development/gitlab.com/biominers/covid19-testing-data/ \
  www/
```


Step 6: Upload plots and table to AWS S3 bucket as static html

```
bash sync_aws.sh
```


Step 7: Exporting to github

We also mirror one file from this repository to github https://github.com/KhazenLab/covid19-data

```
cp ~/Development/gitlab.com/biominers/covid19-testing-data/l1b-altogether/multiple-aggregated_owid_wiki_worldometers_biominers-gitrepo.csv ~/Development/github.com/KhazenLab/covid19-data/covid19-testing-dataset.csv
sed -i "s/biominers/lau/g" covid19-testing-dataset.csv
```


## Shiny apps

The first shiny app was created for the comparison of different countries.

It's committed in this repo at `shinyScripts/compareCountries/compareCountries.R`

Follow instructions in the fil header there.



## Data cleaning

### Part 1: daily tests being negative

We first started doing data cleaning inline in the source code.
Then we started complementing that with a `drop_entries.csv` file in the data repo,
which we read here in the CLI.
The way the `drop_entries.csv` file was built is by opening the `l1b/side-by-side.csv` file in a worksheet editor,
then adding 2 columns to it:

- neg diff: in K2 then drag down:

`=IF(A2=A1, IF(I2-I1<0, "x", ""), "")`

- diff with context: in L21 then drag down:

`=IF(OR(K21="x", K20="x", K19="x", K18="x", K17="x", K16="x", K22="x", K23="x", K24="x", K25="x", K26="x"), "y", "")`

Then I filter the sheet for `diff with context = y`.

Finally I add a `drop row` column which I manually annotate with `d` for rows that I inspect and decide to drop.
Sometimes I drop the row marked with `neg diff=x`, othertimes it's the rows before or after it.
If it's a biominers entry, then I fix the issue in the notion tables.

I finally save it as `drop_entries.csv` for example in the data repo.

For future rows that also need to be dropped, I will need to concatenate the newly done file above with the existing one in the repo.

Update: Adding new batch of data cleaning now.
First, I sort by Location, Date, source.
Then I add a new column `dupe` with `=IF(AND(A2=A1,B2=B1,J2=J1), IF(M2="d",IF(M1<>"d", "important", "candel"), "candel"), "")` in the `drop_entries.csv` file
after I pasted the new rows to be dropped.
Then I needed to mark the row above the dupe with "d" since I need to delete the dupe rows.
Finally I delete all the rows marked with "candel", drop the dupe column, and save to csv.


### Part 2: daily tests less than daily confirmed

#### Part 2a: due to extended data

Saved in `l0/drop_tests_lessthan_confirmed.csv'

Based on `l2/t11c-confirmed+totalTests-historical.csv`, and prioritized in notebook t11d in the section that highlights countries with tests less than confirmed < -1k for example

Procedure
- rename CountryProv to Location, `total_cumul.source` to source
- add `d conf` and `d test` and `test - conf` columns for info
- mark `drop row` manually
- set in S4 formula `=IF(OR(R4="d",R3="d",R2="d",R5="d",R6="d"), "x", "")` to get some context
- drop everything that has empty in the context
- drop unnecessary columns

#### Part 2b: due to linear interpolation mismatch with concave confirmed cases

Data from notion table `Artificial biominers data` (Link: https://www.notion.so/Artificial-biominers-data-dabae264d80b4c8fb9f7751730c05632 )

Check also the section `Data cleaning` (link: https://www.notion.so/Testing-Statistics-Data-Collection-c9ee53ecacfc4eee950fcc466b8dfb72#1ae3ae9a07d04bb891084702d1bbae9a )



## Troubleshooting

### How to deal with l1 error "Found 46 triplets with dips in their cumulative tests."

Here is the current output from l1 while running it on the morning of 05-28

```
# python3 main.py l1-importothers   ~/Development/gitlab.com/biominers/covid19-testing-data/
JHU .. downloading
Download: jhu confirmed global to /tmp/tmpetigqxb6/jhu-global-original.csv: start
Download: jhu confirmed global: end
Download: jhu confirmed usa to /tmp/tmpetigqxb6/jhu-usa-original.csv: start
Download: jhu confirmed usa: end
Download: jhu death global to time_series_covid19_deaths_global.csv: start
Download: jhu death global: end
Download: jhu death usa to time_series_covid19_deaths_US.csv: start
Download: jhu death usa: end
JHU .. done
Download: owid live to /tmp/tmpx2bxxoa6/ourworldindata.org-roser-live.csv: start
Download: owid live: end
Download: cov usa daily to /tmp/tmpx2bxxoa6/covidtracking.com-states_daily.csv: start
Download: cov usa daily: end
Download: cov usa info to /tmp/tmpx2bxxoa6/covidtracking.com-states_info.csv: start
Download: cov usa info: end
Found duplicates in wiki data
Found 46 triplets with dips in their cumulative tests. Here are the top 20:
source            Location       Date total_cumul.source  total_cumul.all  is_wrong  is_approved
1565                Brazil 2020-05-27       worldometers         871839.0      True         True
3300                  Cuba 2020-05-27       worldometers          97003.0      True         True
3359                Cyprus 2020-05-27       worldometers         103814.0      True         True
3482               Czechia 2020-05-27       worldometers         410777.0      True         True
3793    Dominican Republic 2020-05-27       worldometers          70972.0      True         True
4016               Estonia 2020-05-27       worldometers          78085.0      True         True
4218               Finland 2020-05-27       worldometers         171575.0      True         True
4581               Georgia 2020-05-26       worldometers          49093.0      True         True
4655               Germany 2020-05-19               wiki        3595059.0      True        False
4708                 Ghana 2020-05-26       worldometers         197194.0      True         True
5199               Iceland 2020-05-27       worldometers          59087.0      True         True
5795             Indonesia 2020-05-22         owid/roser         168969.0      True        False
6724                 Japan 2020-04-26               wiki         244978.0      True        False
6754                 Japan 2020-05-26       worldometers         435412.0      True         True
6755                 Japan 2020-05-27       worldometers         276170.0      True         True
8301               Myanmar 2020-05-24       worldometers          17875.0      True         True
8515           Netherlands 2020-05-19       worldometers         297347.0      True         True
8895       North Macedonia 2020-05-27       worldometers          25841.0      True         True
9314                  Peru 2020-05-22         owid/roser         112353.0      True        False
9317                  Peru 2020-05-25         owid/roser         123549.0      True        False

      WARNING: Please mark approved (or fix) in l1b-altogether/dropped_outlaws_l1.csv then re-run l1 to re-calculate selected source per country/state/date triplet
            Hint: by default, we mark worldometers drops as approved
```

The download lines are just informative, so there's nothing to do about them ATM.

The line `Found duplicates in wiki data` should be fixed in the wiki data, but we're just ignoring it for now as we are skipping duplicates in the code.

The line `Found 46 triplets with dips in their cumulative tests. Here are the top 20:` is important, and this section describes how to deal with it.

The entries with `is_approved=True` are entries that don't need any action. These are related to the hint below that says "Hint: by default, we mark worldometers drops as approved".

Focusing on the entries with `is_approved=False`, here's a copy of the table above, but filtered for them:

```
Found 46 triplets with dips in their cumulative tests. Here are the top 20:
source            Location       Date total_cumul.source  total_cumul.all  is_wrong  is_approved
4655               Germany 2020-05-19               wiki        3595059.0      True        False
5795             Indonesia 2020-05-22         owid/roser         168969.0      True        False
6724                 Japan 2020-04-26               wiki         244978.0      True        False
9314                  Peru 2020-05-22         owid/roser         112353.0      True        False
9317                  Peru 2020-05-25         owid/roser         123549.0      True        False

      WARNING: Please mark approved (or fix) in l1b-altogether/dropped_outlaws_l1.csv then re-run l1 to re-calculate selected source per country/state/date triplet
            Hint: by default, we mark worldometers drops as approved
```

So we edit the file `l1b-altogether/dropped_outlaws_l1.csv` and look for the entries marked as `is_wrong=True` and examine each of the entries above.

In the case of Germany above, we observe the following from the csv file:

```
Germany,2020-05-10,owid/roser,3177307.0,,
Germany,2020-05-17,owid/roser,3608189.0,,
Germany,2020-05-19,wiki,3595059.0,x,
Germany,2020-05-21,worldometers,3595059.0,,
Germany,2020-05-24,owid/roser,3952971.0,,
Germany,2020-05-28,worldometers,3952971.0,,
```

The first observation is that this dataset is coming from OWID (our top priority source), as well as from wikipedia and worldometers (lower priority sources).
The problem is that the wiki entry on 05-19 (3595059) is smaller than the owid entry on 05-17 (3608189).
To fix this issue, we mark the wiki entry on 05-19 with an x in the last column (i.e. the `is_approved` column) as follows

```
Germany,2020-05-19,wiki,3595059.0,x,x
```

We also notice that the worldometers entry on 05-21 is just a copy of the wiki entry, and hence also has the same problem above.
Therefore, we also mark the worldometers entry with an x in the `is_approved` column as follows

```
Germany,2020-05-21,worldometers,3595059.0,,x
```

With the above 2 approvals, we re-run the `l1` step.
It is not necessary that the `Germany` row on 05-19 disappear from the table of the error message.
The reason is that once we marked the wiki entry above for Germany on 05-19 as skipped,
a re-run of the l1 step will try to fetch that specific date from a lower-priority datasource.
In this case, only worldometers is lower priority than wikipedia, so if worldometers had an entry for Germany on 05-19,
then it would show up in the dataset, and if worldometers had the same problem of being smaller than the owid value on the previous date,
then it would show up again in the table as another row that needs to be skipped.
In reality, worldometers didn't have an entry for Germany on 05-19, so the `Germany` row on 05-19 disappeared from the table of the error message.

Naturally, the other countries would still appear, as we haven't fixed anything about them yet.

Repeating the above procedure for the rest of these countries will lead us to the l1 step running without an error about "dips in cumulative tests".

Note that the l1 step will need to be run multiple times, iterating between running it and editing the `l1b-altogether/dropped_outlaws_l1.csv` file.
The number of times it needs to be run depends on how many new problems show up in the data with the data updates.

Here is the example of Indonesia with an excerpt of the file `l1b-altogether/dropped_outlaws_l1.csv` file

```
Indonesia,2020-05-19,owid/roser,147799.0,,
Indonesia,2020-05-20,owid/roser,154139.0,,
Indonesia,2020-05-21,worldometers,211883.0,,
Indonesia,2020-05-22,owid/roser,168969.0,x,
Indonesia,2020-05-23,owid/roser,176035.0,,
```

In this case, we see that the `is_wrong` field marked with the x is the owid on 05-22.
However, a visual inspection shows that the real problem is in worldometers data on 05-21 in the row before.
In this case, the solution is to mark the worldometers row on 05-21 as approved to skip instead of the owid row on 05-22.


CHANGELOG.md
All entries in reverse chronological order.

Version 1.0 (2020-06-??)

- misc commits undocumented in the changelog (by Halim and Shadi)


Version 0.9 (2020-05-??)

- feat, l4: distinguish p3 slopes fig from p4 chi-squared plots and merge together
- enh, l4: chisquared detail plots to have region
- enh, l4: chisquared + scatter of slopes improved aesthetics
- enh, l4: chisquared added aggregates by world, continents
- enh, l4: p3a chisquared+trends: html responsive
- enh, l4: spin off p3a (simplified) and p3b (detailed) out of p3
- enh, l4: chisquared step now saves the postprocessed csv into the data repository
- enh, l4,p5: scatter plots now contain arrows wtih 14-day history and multiple dates for each 14 days


Version 0.8 (2020-05-13)

- feat, l4: add command l4-plots to generate plots from publication 1
- feat, l2: interpolation in tests by translate + scale from the number of confirmed cases
  - This ensures that the daily number of tests is always greater than or equal to the number of confirmed cases
- feat, l1+l2: split out common code for adding context to boolean column
- feat, l2: incorporate manually-done datacleaning of confirmed cases + add check that no negatives
- enh, l2: improvements to the interpolation after testing in notebook
- feat, l2: drop spikes by easing (from notebook t15b)
- enh, l0: add brazil
- feat, l4: add static html dashboard built with bokeh for postprocessing data
- feat, www: create a new www folder that can be synced to s3 as our internal dashboard
- feat, l4: add spike easing to confirmed cases, make spike easing more strict
  - bugfix: found that some cases were yielding dips, so I added a workaround for those, but it's not a full fix though
- feat, l4: integrate chi-squared dashboard
- bug, l0: san marino needs summation of total cumul fields
- enh, l1: add warning if wiki floats are strings
- enh, l4: plots down to 100 dpi for sharing on the internal dashboard (from 300 dpi for print)


Version 0.7 (2020-05-06)

- feat, l1: start using `drop_entries.csv` for datacleaning
- enh, l0: abort if duplicates in biominers data
- enh, l1: wiki and worldometers had nans, dropped
- bug, l1: covidtracking.com updated urls
- feat, l1: start using `l0/drop_tests_lessthan_confirmed.csv` for datacleaning
- feat, l3: start using interpolation by transformation
- enh, l1: more data cleaning (done in code, but will do in `l0/drop_entries.csv` later, by Halim)
- enh, l0: new countries from notion
- bug, l1: fix expiry setting on cache
- enh, l1: convert exception to warning when dips are found + calculate diff on non-na's
- bug, l2: renames of countries that already exist. Injected a check about it
- feat, l1+l2: CLI outputs l1/drop and l2/drop with enough fields to perform the datacleaning without needing to vimdiff with other csv files
- enh, l0: cambodia table column got renamed
- feat, l1: add skip-download option
- enh, l0: moved some renames from l2 to l0


Version 0.6 (2020-04-29)

- enh, l3: drop the outdated R scripts
- enh, l3: overwrite negative values with 0
- bug, l1: data cleaning of numbers causing negative values
- bug, l1: owid live had duplicates which showed up when added the UID field for data cleaning. Fixed
- feat, l1: add caching for 1 hour for JHU and OWID live files
- bug, l1: drop using the owid frozen version. Check comments in function for more details
- enh, l1: more worldometers country name corrections
- enh, l1: worldometers using non-utf8 encoding in read csv
- enh, l1: worldometers drop extended values


Version 0.5 (2020-04-27)

- 2020-04-27
	- l0 added 5 country/state entries from students
  - compareCountries Initial commit
  - l3 fix
  - l1 fix for nova scotia
  - l3 added 2 new columns to historical and removed some visually weird data
  - l0 added faroe islands
- 2020-04-25
  - l0 handle different columns in UK channel Islands
  - l1 dropped jersey as api is not helpful
  - l0 fix to match JHU name
  - l0 fixed "-"
  - l1 added gov source for jersey island
  - fix indentation
  - l0 handled cyprus
  - l0 added 8 countries
  - l0 changed Northern Macedonia to match dataset
- 2020-04-24
  - l3 fix so that countries with no reporting on daily testing have nas in daily test/million
  - l3 fixed small mistake making total zero equal to na
  - l3 moved lines dealing with cumratio>100 to affect latest as well
- 2020-04-23
  - l3 fix for daily ratio to allow zeros, and for cumulative ratios disallow values >100
- 2020-04-22
  - l3 added conditions dealing with negative ratios and >100 ratios
  - l3 added condition to replace negative ratios with na
  - l3 added conditions to drop negative test/millions and inf ratios
  - l3 added 2 columns to historical (daily ratio and daily tests/million)
- 2020-04-21
  - l0 added isle of man



Version 0.4 (2020-04-21)

- bug, l1: owid had some dupes in india and usa (2 sources for each)
- feat, l1: add owid/freeze to avoid losing poitns that owid drop
- feat, l0: add portugal from owid dropped data
- feat, l0: added 11 countries by students and 4 coutries by HT. Post processed morocco and dominican republic


Version 0.3 (2020-04-20)

- enh, l3: save historical containing ratio positives and tests/mill in separate file
- enh, l1: rename some countries for consistency across sources: US-DC, US-virgin islands, South Korea
- bugfix, l0: for armenia: turns out that the column name called "negative" is actually "total_cumul", so no need for postprocessing anymore
- feat, l1: give our biominers notion tables higher priority than wikipedia
- enh, l0 argentina: use negative+confirmed only for entries in notion where we dont already have total_cumul
- feat, l3: convert from R to py


Version 0.2 (2020-04-16)

- enh, l2: re-enable yemen, samoa, mariana
- feat, l1: jhu deaths also and merge with confirmed
- enh, l2: use jhu data instead of kaggle
- enh, l1: get jhu confirmed cases
- enh, l2: some more details for null lats
- enh, l2: count file now accumulates results

lib/l4_plots.py
# -*- coding: utf-8 -*-
"""t12b-shadi-stats per source again but using l2 data.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/16GiHxAP6SPxWYCugRVF-cV3EVxU2zdY6

Plotting counts per source
"""

from os.path import join
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
import matplotlib
import datetime as dt



class L4Plots:
  """
  Class that reads csv file from l2/historical and generates plots for l4_plots subdir
  """
  def __init__(self):
    self.dt_now = dt.datetime.now().strftime("%Y%m%d_%H%M%S")


  def read_csv(self, dir_gitrepo):
    csv_l2_historical = join(dir_gitrepo, "l2-withConfirmed", "t11c-confirmed+totalTests-historical.csv")
    df_agg = pd.read_csv(csv_l2_historical)
    df_agg.Date = pd.to_datetime(df_agg.Date)

    # Prep plot by using the selected source per triplet
    #This drops the duplicated datapoints (i.e. if a point comes from owid and we get it too, we shouldn't count our point)
    
    
    # for backward compatiability with l1 data
    df_agg["Location"] = df_agg["CountryProv"]
    
    df_agg3 = df_agg[["Location", "Date", "total_cumul.source"]].copy()
    df_agg3 = df_agg3.groupby(["Date", "total_cumul.source"]).count()
    df_agg3 = df_agg3.unstack("total_cumul.source")
    df_agg3.columns = df_agg3.columns.droplevel()
    
    ## show till date - 1. TODO drop this later
    #df_agg3.reset_index(inplace=True)
    #dt_max = df_agg3.Date.max() - pd.to_timedelta(1,'d')
    #df_agg3 = df_agg3[df_agg3.Date <= dt_max]
    #df_agg3.set_index("Date", inplace=True)
    
    # merge roser and ortiz
    df_agg3["owid/roser"] = df_agg3["owid/roser"].fillna(0) + df_agg3["owid/ortiz"].fillna(0)
    del df_agg3["owid/ortiz"]
    
    df_agg3 = df_agg3[["owid/roser", "covidtracking.com", "wiki", "worldometers", "biominers"]]
    
    df_agg3 = df_agg3.rename(columns={
        "owid/roser": "OWID",
        "covidtracking.com": "CTP",
        "wiki": "Wikipedia",
        "worldometers": "Worldometers",
        "biominers": "LAU manual subset"
    })

    self.df_agg3 = df_agg3


  def prep_plots(self):
    # http://stackoverflow.com/questions/3899980/ddg#3900167
    font = {'family' : 'DejaVu Sans',
            'weight' : 'normal',
            'size'   : 22}
    
    matplotlib.rc('font', **font)


  def plot_line(self, dir_plot_destination):
    # line plot
    df_agg3 = self.df_agg3

    plt.figure(figsize=(15,8))
    filled_markers = ('o', 'v', '^', '<', '>', 's', '8', 'p', '*', 'h', 'H', 'D', 'd', 'P', 'X')
    ax=sns.lineplot(data=df_agg3, markers=filled_markers)
    
    # 737530 is for 2020-04-15
    last_date = (df_agg3.index.max() - pd.to_datetime(dt.date(2020,4,15))).days
    extraticks = [737530+last_date]
    plt.xticks(list(plt.xticks()[0]) + extraticks)
    
    ax.set_ylabel("Number of daily country/state pairs")
    plt.xticks(rotation=45)
    plt.grid(alpha=.1)
    plt.ylim(0)
    plt.xlim(right=df_agg3.index.max())
    
    #fn_line = join(dir_plot_destination, 't12b-plotSourcesOverTime-lines-v%s.png'%self.dt_now)
    fn_line = join(dir_plot_destination, 't12b-plotSourcesOverTime-lines.png')
    #plt.savefig(fn_line, dpi = 300, bbox_inches="tight")
    plt.savefig(fn_line, dpi = 100, bbox_inches="tight")
    print(f"Saved to {fn_line} (at 100 dpi)")


  def plot_stacked(self, dir_plot_destination):
    # https://python-graph-gallery.com/251-stacked-area-chart-with-seaborn-style/
    # stacked area plot
    df_agg3 = self.df_agg3
    
    # Data
    x = df_agg3.index
    y=df_agg3.fillna(0).transpose().values
     
    # Plot
    plt.figure(figsize=(15,8))
    ax = plt.stackplot(x, y, labels=df_agg3.columns)
    plt.xlim(x.min(), x.max())
    plt.legend(loc='upper left')
    plt.ylabel("Countries/states with historical data")
    plt.xlabel("Date")
    
    # Add extra tick for May 3 since the automatic ticks stop at May 1
    #last_date = (df_agg3.index.max() - pd.to_datetime(dt.date(2020,4,15))).days
    ## 737530 represents 04-15
    #extraticks = [737530+last_date] # append last date
    ##extraticks = []
    #existingticks = list(plt.xticks()[0])
    ## drop 2020-05-01 because it's too close to 05-03
    #existingticks = existingticks[0:(len(existingticks)-1)]
    #plt.xticks(existingticks + extraticks)
    
    plt.xticks(rotation=45, ha="right")
    
    plt.grid(alpha=.2)
    
    # FIXME set an xlim on 2020-05-01 for lancet correspondence, but that got rejected
    #plt.xlim(right=dt.date(2020,5,1))
    
    # png for doc, and jpg for attachment to submission
    #fn_st_png = join(dir_plot_destination, 't12b-plotSourcesOverTime-stacked-v%s.png'%self.dt_now)
    fn_st_png = join(dir_plot_destination, 't12b-plotSourcesOverTime-stacked.png')
    #plt.savefig(fn_st_png, dpi = 300, bbox_inches="tight")
    plt.savefig(fn_st_png, dpi = 100, bbox_inches="tight")
    print(f"Saved to {fn_st_png} (at 100 dpi)")

    # disabled because it doesn't run on my laptop, but works fine on colab
    #fn_st_jpg = join(dir_plot_destination, 't12b-plotSourcesOverTime-stacked-v%s.jpg'%self.dt_now)
    #plt.savefig(fn_st_jpg, dpi = 300, bbox_inches="tight")

############################################################

from bokeh.layouts import gridplot
from bokeh.models import CDSView, ColumnDataSource, GroupFilter
from bokeh.models import CustomJS
from bokeh.plotting import figure, show
import pandas as pd
from bokeh.plotting import output_file
from bokeh.layouts import column
from bokeh.plotting import save


class PostprocessingDashboard:
  """Notebook t16a"""

  def read_csv(self, dir_gitrepo):
    dir_l2_withConf = join(dir_gitrepo, "l2-withConfirmed")
    df = pd.read_csv(join(dir_l2_withConf, "interpolated_by_transformation.csv"))
    df["Date"] = pd.to_datetime(df.Date)
    
    # replace nans: https://github.com/bokeh/bokeh/issues/4472#issuecomment-225676759
    for key in df:
        df[key] = ['NaN' if pd.isnull(value) else value for value in df[key]]

    self.df = df
        

  def to_html(self, dir_plot_destination):
    fn_dest = join(dir_plot_destination, "t16a-postprocessing_dashboard.html")
    output_file(fn_dest)
    
    source = ColumnDataSource(self.df)
    
    init_group = 'Lebanon'
    gf = GroupFilter(column_name='CountryProv', group=init_group)
    view1 = CDSView(source=source, filters=[gf])
    
    plot_size_and_tools = {'plot_height': 300, 'plot_width': 300,
                            'tools':['box_select', 'reset', 'help', 'box_zoom'],
                          'x_axis_type': 'datetime'}
    
    p_cc_ori = figure(title="Confirmed cases: original", **plot_size_and_tools)
    c_cc_ori = p_cc_ori.circle(x='Date', y='ConfirmedCases', source=source, color='black', view=view1)
    
    p_cc_cl = figure(title="Confirmed cases: eased spikes", **plot_size_and_tools)
    c_cc_cl = p_cc_cl.circle(x='Date', y='cases_cumulClean', source=source, color='red', view=view1)
 
    p_cto = figure(title="Cumulative Tests: original", **plot_size_and_tools)
    c_cto = p_cto.circle(x='Date', y='total_cumul.all', source=source, color='black', view=view1)
    
    p_ci = figure(title="Cumulative Tests: interpolated", **plot_size_and_tools)
    c_ci = p_ci.circle(x='Date', y='tests_cumulInterpolated', source=source, view=view1, color='red')
    
    p_cns = figure(title="Cumulative Tests: eased spikes", **plot_size_and_tools)
    c_cns = p_cns.circle(x='Date', y='tests_cumulNoSpike', source=source, view=view1, color='red')
    
    g = gridplot([[p_cc_ori, p_cc_cl], [p_cto, p_ci, p_cns]])
    
    # from https://docs.bokeh.org/en/latest/docs/user_guide/interaction/widgets.html#select
    callback = CustomJS(args=dict(vf=c_cc_ori.view.filters[0], source=source), code="""
    console.log(vf.group);
    console.log(cb_obj.value);
        vf.group = cb_obj.value;
        source.change.emit();
    """)
    from bokeh.models import Select
    select = Select(title="Country/State:", value=init_group, options=list(self.df.CountryProv.unique()))
    select.js_on_change('value', callback)
    
    layout = column(select, g)
    save(layout)
    print(f"Saved to {fn_dest}")

######################################

class ChisquaredDashboard:
  """Notebook t11d
     Bokeh dashboard code same as postprocessed dashboard above
  """

  def read_csv(self, dir_gitrepo):
    import click
    click.secho("As of 2020-05-13, remember that the l4/chi-squared dashboard requires running notebook t11d and downloading the resultant csv", fg="yellow")
    dir_l4 = join(dir_gitrepo, "l4-analysis")
    df = pd.read_csv(join(dir_l4, "t11d-chisquared-history-v20200512.csv"))
    df.rename(columns={"case_d2": "case_mvsum07", "control_d2": "control_mvsum07"}, inplace=True)
    df["Date"] = pd.to_datetime(df.Date)
    df["case_ma07"] = df["case_mvsum07"]/7
    df["control_ma07"] = df["control_mvsum07"]/7
    df["tests_ma07"] = df.case_ma07 + df.control_ma07
    df["threshold_min"] = df.threshold_min.round()
    df["threshold_max"] = df.threshold_max.round()
    df["case_ma07_eps"] = df.case_ma07*1.03
    df["tests_ma07_eps"] = df.tests_ma07*1.03
    df["threshold_min_eps"] = df.threshold_min*1.03
    df["threshold_max_eps"] = df.threshold_max*1.03
    df["ratio_daily"] = df.daily_conf / df.daily_tests * 100
    df["ratio_ma07"] = df.case_ma07 / df.tests_ma07 * 100
    df["case_detrended"] = df.case_mvsum07 - (df.threshold_min+df.threshold_max)/2

    # columns subset
    df = df[["CountryProv","Date",
             "daily_conf", "case_ma07", "case_mvsum07", "case_ma07_eps",
             "daily_tests", "tests_ma07", "tests_ma07_eps",
             #"control_mvsum07",
             "threshold_min", "threshold_max",
             "threshold_min_eps", "threshold_max_eps",
             "ratio_daily", "ratio_ma07",
             "case_detrended"
             ]]

    # replace nans: https://github.com/bokeh/bokeh/issues/4472#issuecomment-225676759
    #df = df.fillna("NaN")
    #for key in df:
    #    print(f"Replace {key} nans")
    #    df[key] = ['NaN' if pd.isnull(value) else value for value in df[key]]

    self.df = df
        

  def to_html(self, dir_plot_destination):
    fn_dest = join(dir_plot_destination, "t11d-chisquared_dashboard.html")
    output_file(fn_dest)
    
    source = ColumnDataSource(self.df)
    
    init_group = 'Lebanon'
    gf = GroupFilter(column_name='CountryProv', group=init_group)
    view1 = CDSView(source=source, filters=[gf])
    
    plot_size_and_tools = {'plot_height': 300, 'plot_width': 600,
                            'tools':['box_select', 'reset', 'help', 'box_zoom'],
                          'x_axis_type': 'datetime'}
    
    # FIXME couldnt do p_a1.line below, so using hack of varea
    p_a1 = figure(title="Confirmed cases (daily vs 7-day moving avg)", **plot_size_and_tools)
    c_a1a = p_a1.circle(x='Date', y='daily_conf', source=source, color='black', view=view1)
    c_a1b = p_a1.varea(x='Date', y1='case_ma07', y2='case_ma07_eps', source=source, color='red', view=view1)
   
    p_a2 = figure(title="Total tests (daily vs 7-day moving avg)", **plot_size_and_tools)
    c_a2a = p_a2.circle(x='Date', y='daily_tests', source=source, color='black', view=view1)
    c_a2b = p_a2.varea(x='Date', y1='tests_ma07', y2="tests_ma07_eps", source=source, color='red', view=view1)

    p_b1 = figure(title="Confirmed and thresholds (7 vs 14-day sum, below threshold: good, above: bad, within: ok)", **plot_size_and_tools)
    c_b1b = p_b1.varea(x='Date', y1='threshold_min_eps', y2='threshold_max_eps', source=source, color='grey', view=view1)
    c_b1a = p_b1.circle(x='Date', y='case_mvsum07', source=source, color='red', view=view1)

    p_b2 = figure(title="Detrended cases (7-day ma, cases minus thresholds, negative: good, positive: bad)", **plot_size_and_tools)
    c_b2a = p_b2.circle(x='Date', y='case_detrended', source=source, color='green', view=view1)

    p_c1 = figure(title="Ratio case/total (daily)", **plot_size_and_tools)
    c_c1a = p_c1.circle(x='Date', y='ratio_daily', source=source, color='blue', view=view1)

    p_c2 = figure(title="Ratio case/total (7-day ma)", **plot_size_and_tools)
    c_c2a = p_c2.circle(x='Date', y='ratio_ma07', source=source, color='blue', view=view1)

    g = gridplot([[p_a1, p_a2], [p_b1, p_b2], [p_c1, p_c2]])
    
    # from https://docs.bokeh.org/en/latest/docs/user_guide/interaction/widgets.html#select
    callback = CustomJS(args=dict(vf=c_a1a.view.filters[0], source=source), code="""
    console.log(vf.group);
    console.log(cb_obj.value);
        vf.group = cb_obj.value;
        source.change.emit();
    """)
    from bokeh.models import Select
    select = Select(title="Country/State:", value=init_group, options=list(self.df.CountryProv.unique()))
    select.js_on_change('value', callback)
    
    layout = column(select, g)
    save(layout)
    print(f"Saved to {fn_dest}")
